src/ips.rs

```diff
@@ -0,0 +1,431 @@
+// Intelligent Peer Sharing (IPS) module
+// Selection is based on the "beauty" contest of the nodes - each node is evaluated based on its
+// degree, betweenness, closeness and eigenvector centrality. Then, if requested ranking is
+// updated with location factor. Each factor has its own weight that is used to determine
+// factor's importance to the calculation of the final ranking. That gives possibility
+// to test different approaches to the selection of the peers, without re-compiling the code.
+// Weights are defined in the configuration file.
+// When the ranking is calculated, the peers are selected based on the ranking. The number of
+// peers can be changed and it is defined in the configuration file. Algorithm is constructed
+// as step by step process, so it is easy to add new steps or change the order of the steps.
+// Especially, there could be a need to add some modifiers to the ranking.
+
+use crate::config::IPSConfiguration;
+use crate::{CrunchyState, Node};
+use geoutils::Location;
+use serde::{Deserialize, Serialize};
+use spectre::edge::Edge;
+use spectre::graph::{AGraph, Graph};
+use std::collections::HashMap;
+use std::collections::{HashSet, VecDeque};
+use std::net::IpAddr;
+use std::str::FromStr;
+
+/// Intelligent Peer Sharing (IPS) module structure
+#[derive(Default, Clone)]
+pub struct Ips {
+    config: IPSConfiguration,
+    degree_factors: NormalizationFactors,
+    betweenness_factors: NormalizationFactors,
+    closeness_factors: NormalizationFactors,
+    eigenvector_factors: NormalizationFactors,
+}
+
+/// Peer list structure containing peer list for each node
+#[derive(Clone, Serialize, Deserialize)]
+pub struct PeerList {
+    /// IP address of the node
+    pub peer: IpAddr,
+    /// List of peers for the node
+    pub peer_list: Vec<IpAddr>,
```

Comment:
```suggestion
pub struct Peer {
    /// IP address of the node
    pub ip: IpAddr,
    /// List of peers for the node
    pub list: Vec<IpAddr>,
```
Or maybe `pub struct PeerInfo {`

Reply:
Done.